### PR TITLE
Reduce number of times optimizers are instantiated with FSDP

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Follow up to comment here: https://github.com/PyTorchLightning/pytorch-lightning/pull/11952#discussion_r820060750

Currently we are calling `LightningModule.configure_optimizers` 3 times in the FSDP strategy 😞
We only need to do this one, after the model is wrapped with FSDP inside of `model_sharded_context` 

## Before submitting
No

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/strategies/fully_sharded.py [hunk 1]
```diff
@@ -135,16 +135,11 @@ def setup_distributed(self) -> None:
...
         self.accelerator.setup(trainer)
-        self.setup_optimizers(trainer)
-        self.setup_precision_plugin()
-        optimizers_to_device(self.optimizers, self.root_device)
 
         if self._layer_sync:
             self.model = self._layer_sync.apply(self.model)
 
         self.configure_ddp()
-        self.barrier()
```
Comment (awaelchli) [inline pytorch_lightning/strategies/fully_sharded.py:146]: Are you sure the barrier can be dropped? It is there for a reason, we should probably keep it.

File: pytorch_lightning/strategies/fully_sharded.py [hunk 2]
```diff
@@ -171,19 +166,19 @@ def wrap_policy(*args, **kwargs):
             yield
 
         log.detail(f"{self.__class__.__name__}: exiting model_sharded_context.")
+        self.setup_optimizers(self.lightning_module.trainer)
+        optimizers_to_device(self.optimizers, self.root_device)
+        self.setup_precision_plugin()
```
Comment (awaelchli) [inline pytorch_lightning/strategies/fully_sharded.py:171]: why can it not be done in setup? setup() is mean to do exactly that, and setup gets called after the model is wrapped (see trainer.py). I highly recommend to keep it in setup()

Comment (ananthsub) [reply pytorch_lightning/strategies/fully_sharded.py:171]: Yes, I can change that. I needed to double check the ordering around wrapping FSDP vs move to device. 

But looking again at the trainer code, why does the Trainer need to explicitly call configure_sharded_model. Could we move that to the Strategy's setup instead?

File: pytorch_lightning/strategies/fully_sharded.py [hunk 3]
```diff
@@ -135,16 +135,15 @@ def setup_distributed(self) -> None:
...
-        optimizers_to_device(self.optimizers, self.root_device)
 
         if self._layer_sync:
             self.model = self._layer_sync.apply(self.model)
 
         self.configure_ddp()
         self.barrier()
         self.setup_optimizers(trainer)
+        optimizers_to_device(self.optimizers, self.root_device)
+        self.setup_precision_plugin()
```
Comment (four4fish) [inline pytorch_lightning/strategies/fully_sharded.py:146]: Does the order for these two matter? I see inconsistency across different strategies

**Review by four4fish — APPROVED** [review]

File: pytorch_lightning/strategies/fully_sharded.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/strategies/fully_sharded.py:146]: i prefer updating the optimizers first. setup_precision_plugin simply passes the model, optimizers, and lr_schedulers to the precision plugin to view

**Review by awaelchli — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by Borda — APPROVED** [review]
